lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,298 @@
+  Region &funcReg = funcOp.getRegion();
+  DenseMap<Block *, llvm::SmallVector<Block *, 4>> control_deps_map;
+
+  // Initialize the all_control_deps_maps by creating an entry for every block
+  // constituting the region
+  for (Block &block : funcReg.getBlocks()) {
+    SmallVector<Block *, 4> deps;
+    control_deps_map.insert(std::make_pair(&block, deps));
+  }
+
+  // functions made up of 1 basic block do not have any dependencies
+  if (funcReg.hasOneBlock())
+    return;
+
+  // Get information about post-dominance
+  PostDominanceInfo postDomInfo;
+  // Get the post-dominance tree
+  llvm::DominatorTreeBase<Block, true> &postDomTree =
+      postDomInfo.getDomTree(&funcReg);
+
+  // Loop over the control flow edges connnecting the different blocks of this
+  // region
+  for (Block &block : funcReg.getBlocks()) {
+    for (Block *block_succ : block.getSuccessors()) {
+      if (!postDomInfo.properlyPostDominates(block_succ, &block)) {
+        Block *least_common_anc =
+            postDomInfo.findNearestCommonDominator(block_succ, &block);
+
+        if (least_common_anc == &block)
+          // Loop case
+          control_deps_map[&block].push_back(&block);
+
+        // All nodes in the post-dominator tree on the path from the
+        // "least_common_anc" to "block_succ" (including "block_succ") should
+        // be control dependent on "block"
+        // Easy case of block_succ
+        control_deps_map[block_succ].push_back(&block);
+        // Enumerate all paths between "least_common_anc" and "block_succ"
+        SmallVector<PathInDomTree, 4> allPathsFromLeastCommonAncToBlockSucc;
+        enumeratePathsInPostDomTree(least_common_anc, block_succ, &funcReg,
+                                    &postDomTree,
+                                    &allPathsFromLeastCommonAncToBlockSucc);
+
+        for (PathInDomTree path : allPathsFromLeastCommonAncToBlockSucc)
+          for (DominanceInfoNode *domInfo : path) {
+            Block *blockInPath = domInfo->getBlock();
+
+            // Skip the nodes that we have already taken care of above
+            if (blockInPath == least_common_anc || blockInPath == &block ||
+                blockInPath == block_succ)
+              continue;
+
+            // Add block to the control dependencies only if it is not already
+            // there
+            if (std::find(control_deps_map[blockInPath].begin(),
+                          control_deps_map[blockInPath].end(),
+                          &block) == control_deps_map[blockInPath].end())
+              control_deps_map[blockInPath].push_back(&block);
+          }
+      }
+    }
+  }
+
+  // Up to this point, we have correct direct dependencies that do not include
+  // nested dependencies (i.e., dependencies of a block's dependencies) call the
+  // following function to include nested dependencies
+  addDepsOfDeps(funcOp, control_deps_map);
+  all_control_deps_maps.emplace_back(control_deps_map);
```

Comment:
This copies the entire map, you can instead do the following at the beginning of the method.

```cpp
DenseMap<Block *, llvm::SmallVector<Block *, 4>>& control_deps_map = all_control_deps_maps.emplace_back();
```

This adds an empty map to the vector and returns you a reference to it you can directly modify.

Reply:
No longer relevant after I changed the data structure holding the control dependencies, as per your earlier suggestion.